framework/validators/IpValidator.php

```diff
@@ -0,0 +1,555 @@
+     * expanded to `2001:0db8:0000:0000:0000:0000:0000:0001`
+     *
+     * @param string $ip the original IPv6
+     * @return string the expanded IPv6
+     */
+    public static function expand6($ip)
+    {
+        $hex = unpack("H*hex", inet_pton($ip));
+        return substr(preg_replace("/([a-f0-9]{4})/i", "$1:", $hex['hex']), 0, -1);
+    }
+
+    /**
+     * The method checks, if the IP address is allowed according to
+     * the [[allow]] and [[deny]] lists, and the [[order]] property.
+     *
+     * @param string $ip
+     * @param integer $cidr
+     * @return boolean
+     * @see order
+     */
+    public function isAllowed($ip, $cidr)
+    {
+        $denied = false;
+        $allowed = true;
+        if (!empty($this->deny) && $this->inRange($ip, $cidr, $this->deny)) {
+            $denied = true;
+        }
+        if (!empty($this->allow) && !$this->inRange($ip, $cidr, $this->allow)) {
+            $allowed = false;
+        }
+
+        return $this->order === self::ORDER_DENY_ALLOW ? (!$denied && $allowed) : (!$denied || $allowed);
+    }
+
+    /**
+     * Validates IPv4 address
+     *
+     * @param string $value
+     * @return boolean
+     */
+    public function validate4($value)
+    {
+        return preg_match($this->ipv4Pattern, $value) !== 0;
+    }
+
+    /**
+     * Validates IPv6 address
+     *
+     * @param string $value
+     * @return boolean
+     */
+    public function validate6($value)
+    {
+        return preg_match($this->ipv6Pattern, $value) !== 0;
+    }
+
+    /**
+     * Gets the IP version
+     *
+     * @param string $ip
+     * @return integer
+     */
+    public static function getIpVersion($ip)
+    {
+        return strpos($ip, ":") === false ? 4 : 6;
+    }
+
+    /**
+     * Used to get the Regexp pattern for initial IP address parsing
+     * @return string
+     */
+    public function getIpParsePattern()
+    {
+        $negationChar = is_string($this->negationChar) ? $this->negationChar : static::DEFAULT_NEGATION_CHAR;
+        return '/^(' . $negationChar . '?)(.+?)(\/(\d+))?$/';
+    }
+
+    /**
+     * Getter for [[_negationChar]]
+     *
+     * @return string
+     * @see _negationChar
+     */
+    public function getNegationChar()
+    {
+        return $this->_negationChar;
+    }
+
+    /**
+     * Setter for [[_negationChar]]
+     *
+     * @param $negationChar
+     * @see _negationChar
+     */
+    public function setNegationChar($negationChar)
+    {
+        if (is_string($this->negationChar)) {
+            $negationChar = preg_quote($this->_negationChar, '/');
+        }
+        $this->_negationChar = $negationChar;
+    }
+
+
+    /**
+     * Checks whether the IP is in subnet ranges
+     *
+     * @param string $ip an IPv4 or IPv6 address
+     * @param integer $cidr
+     * @param string|array $ranges allowed subnets in CIDR format e.g. `10.0.0.0/8` or `2001:af::/64`
+     * @return bool
+     */
+    public function inRange($ip, $cidr, $ranges)
```

Comment:
Well, private suits it better for now since it's not inteneded to be exposed as far as I can tell.